google/cloud/storage/retry.py

```diff
@@ -22,6 +22,7 @@
 
 
 _RETRYABLE_TYPES = (
+    ConnectionError,
```

Comment:
ConnectionError has a bunch of subclasses: https://docs.python.org/3/library/exceptions.html#ConnectionError . Just wanted to confirm that we consider all of these retryable?

Also, is there a test where we should add this as a test case?

Reply:
One or more tests should be added to the `Test_should_retry` class in `tests/unit/test_retry.py` (see the [BigQuery PR](https://github.com/googleapis/python-bigquery/pull/571/files) for examples).

Also, that PR makes `requests.exceptions.ConnectionError` an explicit retryable type.  The `requests` version does *not* derive from the stdlib type:

```python
$ .nox/unit-3-9/bin/python
Python 3.9.0 (default, Apr 20 2021, 11:50:03) 
[GCC 9.3.0] on linux
>>> import requests.exceptions
>>> requests.exceptions.ConnectionError
<class 'requests.exceptions.ConnectionError'>
>>> requests.exceptions.ConnectionError.__mro__
(<class 'requests.exceptions.ConnectionError'>, <class 'requests.exceptions.RequestException'>, <class 'OSError'>, <class 'Exception'>, <class 'BaseException'>, <class 'object'>)
```

Reply:
Note that `ConnectionError` is not a stdlib type in Python 2.7:

```python
$ .nox/unit-2-7/bin/python
Python 2.7.17 (default, Apr 20 2021, 14:27:04) 
[GCC 9.3.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> ConnectionError
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'ConnectionError' is not defined
```

We need to do something like:

```python
try:
    _RETRYABLE_STDLIB_TYPES = {ConnectionError,)
except NameError:
    _RETRYABLE_STDLIB_TYPES = {}

```
and then, below:
```python
RETRYABLE_TYPES = _RETRYABLE_STDLIB_TYPES + (
    api_exceptions.TooManyRequests,  # 429
    ...
)

Reply:
@tritone @tseaver Thanks for reviewing! Looking at the python3 docs and [previous discussion in the Big Query](https://github.com/googleapis/python-bigquery/issues/434), ConnectionError and its subclasses are retryable. From users feedback, this error occurs when there are multiple Cloud Functions running and/or subsequent reinvocation of Cloud Functions. The [retry test](https://github.com/googleapis/python-storage/blob/3fb5086d54ebe487e54cb5a254c0899c91faaf8e/tests/unit/test_retry.py#L40) is set up with a loop that tests each default retry. I'll double check to make the changes and add one or more specific tests. Thanks!

Reply:
@tseaver I noticed that `ConnectionError` is not a stdlib type in Python 2.7 upon submission. Changed PR to draft to implement changes. This is exactly what I need. Thanks so much for the pointers!